### PR TITLE
fix(oidcclient):for issue 734

### LINKED ISSUE #734: using hash router, getting the parameters from the url is incorrect
Situation:
When the router is set to hashmode the parameters are not correctly parsed from the url.
and you get the error that the state is not found.

Analysing:
- UrlUtils.readParams expects an unhashed routed url

easiest way yo fix: 
const unHashUrl = url.replace('#/', '');
const response = new SigninResponse(UrlUtils.readParams(unHashUrl, this.settings.response_mode));

and 

const unHashUrl = url.replace('#/', '');
const response = new SignoutResponse(UrlUtils.readParams(unHashUrl, this.settings.response_mode));

### PR DESCRIPTION
**Pull request by frontendplace**
Closes/fixes #734

### Checklist

- [ ] This PR makes changes to the public API 
- [x] I have included links for closing relevant issue numbers

### CODE REVIEW

File: src/utils/UrlUtils.ts [hunk 1]
```diff
@@ -7,6 +7,15 @@
 export class UrlUtils {
     public static readParams(url: string, responseMode: "query" | "fragment" = "query"): URLSearchParams {
         if (!url) throw new TypeError("Invalid URL");
+
+        // if we're doing query, then strip off hash fragment before we parse
+        if (responseMode === "query") {
+            const hashIdx = url.indexOf("#");
+            if (hashIdx >= 0) {
+                url = url.substr(hashIdx + 1);
```
Comment (pamapa) [inline src/utils/UrlUtils.ts:15]: please replace `substr` with not deprecated `substring` version

File: src/utils/UrlUtils.ts [hunk 2]
```diff
@@ -7,6 +7,15 @@
...
+            }
+        }
+
         const parsedUrl = new URL(url, window.location.origin);
```
Comment (pamapa) [inline src/utils/UrlUtils.ts:19]: What about the baseUrl `window.location.origin`? Do we need to strip here as well?

**Comment by pamapa** [comment]
Thanks for taking care!

File: src/utils/UrlUtils.ts (hunk 2 above)
Comment (frontendplace) [reply src/utils/UrlUtils.ts:19]: no the origin doesn't have a hash

File: src/utils/UrlUtils.ts (hunk 1 above)
Comment (frontendplace) [reply src/utils/UrlUtils.ts]: done

**Comment by pamapa** [comment]
@longsleep Can you please have a look too?

File: src/utils/UrlUtils.test.ts [hunk 3]
```diff
@@ -34,6 +34,16 @@ describe("UrlUtils", () => {
             expect(resultObj).toHaveProperty("foo", "test");
         });
 
+        it("should return query params when hash routes are used", () => {
+            // act
+            const result = UrlUtils.readParams("http://app/#/path?code=foo&status=bar");
```
Comment (longsleep) [inline src/utils/UrlUtils.test.ts:39]: This is not the correct way to define urls with query parameter and imo it is not our job to deal with this..

Everything after the `#` is part of the hash and that includes the `?` etc. Means in technical terms this URL has no query.

Only the application specific hash based router knows about this and thus it must do URL parsing before our fallback is run.

To fix the test, please put the ? and it's values before the # so we ensure that this works (removing the change from below).

**Review by longsleep — CHANGES_REQUESTED** [review]
This can cause undesired results as imo the assumptions which are base for this pr are wrong. Should probably discussed further in the associated issue

**Comment by frontendplace** [comment]
Ill remove the PR. 
I can see that I can pass a modified url in angular when calling the signinRedirectCallback function

File: src/utils/UrlUtils.test.ts (hunk 3 above)
Comment (frontendplace) [reply src/utils/UrlUtils.test.ts:39]: De PR can be deleted I see I can fix this in our angular Code by passing a modified version of the default window.location.href

**Pull request closed without merging**